docs/internal/constrained-memory.md

```diff
@@ -0,0 +1,175 @@
+
+### Scheduling a behaviour on low-memory conditions
+
+In theory, we could define a closure to handle low-memory conditions and have the runtime schedule it on a designated cown when memory is exhausted.
+This is very difficult to make work correctly because the message may come behind others and may not run immediately.
+It also has the same problem as pre-allocated `when` blocks: out-of-memory conditions are global and a `cown` receiving the behaviour may not be able to address them.
+Even if scheduled to run on a cown with a lot of state, it my be scheduled while that cown is running a behaviour that allocates a lot of memory and so not be run until after the cown has hit unrecoverable low-memory conditions.
+
+Possible solutions
+------------------
+
+There are a few solutions that might be useful.
+
+### Pre-allocate regions
+
+The pattern of pre-allocating objects for a specific task is very easy to map to Verona with a new region kind.
+Existing region implementations will call `snmalloc` to allocate either individual objects or chunks, on demand.
+It would be easy to extend this to provide a region that pre-allocated sufficient memory for all allocations.
+The constructor for the region would need to throw an exception on failure allowing code to handle potential failure of a set of allocations in a single point.
+
+Currently, Verona lacks an equivalent of C's `sizeof` operator and so does not have a way of specifying the amount of memory required in terms of something relevant to the programmer.
+It might be possible for the constructor for such a region to be a generic that took a type list.
+For example:
+
+```verona
+try
+{
+  var region = PrereservedRegion[Array[SomeCls, 42], SomeBigCls, Array[SomeSmallCls, 1000]];
+  allocateABunchOfStuff(region);
+}
+catch
+{
+  { OutOfMemory => /* report failure here */ }
+}
+```
+
+### Low-memory noticeboards
+
+XNU, Android, and Windows all provide some form of low-memory notifications.
+The most useful forms of these are when memory is fairly low but allocation remains possible.
+These can be used by software to avoid allocation or free existing objects when memory is low but not yet exhausted.
+These are particularly useful in garbage-collected systems where it's often beneficial to avoid running the GC until the system starts to experience some memory pressure.
+
+In a Verona world, traced regions are only collected in response to explicit invocation of the collector.
+A cown may wish to avoid tracing until memory starts to be constrained.
+Signalling a noticeboard and scheduling the cowns that have requested notifications does not require memory allocation and so providing a low-memory noticeboard would allow a cown to perform some cleanup when memory is low, without requiring any additional allocation.
```

Comment:
But again, as above, this message may arrive much later, after additional (de)allocation... so it seems like this may be more of a "memory was low in the past" vs. "memory is low now" thing, which seems less useful?

Reply:
Kind of.  I'd missed that when @mjp41 fixed my code, we'd lost the property that notifications were unordered with respect to behaviours, though even with that the cown would still need to get to the front of the scheduler queue before it could handle the notification.  This is why I think this is a lot more useful as a 'memory is a bit low' notification than an aggressive 'you need to free things now!' message.

That said, notifications are both edge triggered and coalescing.  If you register to receive this notification then your notification callback will receive the value of the noticeboard *at the time when the callback is invoked*.  This may lead to some confusion, where you receive a low-memory notification telling you that memory is not low, but as long as your notification callback guards that properly then it can very quickly discard notifications that memory was low in the past.